### ISSUE TITLE
Question about external tutorial/example link

### ISSUE THREAD

**Issue by ilos-vigil** [amaiya/ktrain#406]
First of all, thank you for writing this library. Some time ago i finished my thesis/[skripsi](https://id.wikipedia.org/wiki/Skripsi) for Bachelor degree. The thesis is about evaluting ktrain on text domain where i released some of code/jupyter notebook at https://github.com/ilos-vigil/ktrain-assessment-study.

My question is, would you like to include my repository as tutorial/example on ktrain's README.md?

**Comment by ilos-vigil** [comment]
The wording looks fine to me. On a side note, would you like to receive possible enhancement for ktrain? If yes, i'll make few issues.

**Comment by amaiya** [comment]
Great - if you'd like, we can discuss the proposed enhancements as part of this thread before making separate issues or PRs for them.

**Comment by ilos-vigil** [comment]
Sure, i'll compile list of possible enhancements and double check first.

**Comment by ilos-vigil** [comment]
Here's the list of possible enhancement that i deemed realistic.

## 1. Naming inconsistency

When splitting data, ktrain refer splitted data either as train/test or train/val. For example,

* Function [`texts_from_folder`](https://github.com/amaiya/ktrain/blob/v0.28.3/ktrain/text/data.py#L13-L21) use "train/test"

```py
def texts_from_folder(datadir, classes=None, 
                      max_features=MAX_FEATURES, maxlen=MAXLEN,
                      ngram_range=1,
                      train_test_names=['train', 'test'],
``` 

* Function [`texts_from_csv`](https://github.com/amaiya/ktrain/blob/v0.28.3/ktrain/text/data.py#L125-L128) "train/val"

```py
def texts_from_csv(train_filepath, 
                   text_column,
                   label_columns = [],
                   val_filepath=None,
```

## 2. Adding more vision model

At this moment, ktrain only offer [8](https://github.com/amaiya/ktrain/blob/v0.28.3/ktrain/vision/models.py#L18-L26) different vision model. I suggest to add few more models from [`tf.keras.applications`](https://www.tensorflow.org/api_docs/python/tf/keras/applications). My recommendation are,

* MobileNetV3
* EfficientNet

## 3. Additional option during preprocessing text

Add few additional parameter for [standard text processing](https://github.com/amaiya/ktrain/blob/v0.28.3/ktrain/text/preprocessor.py#L542).

1. Let user add custom preprocessing function before [extracting vocabulary](https://github.com/amaiya/ktrain/blob/v0.28.3/ktrain/text/preprocessor.py#L612-L616).

```py
import re
import ktrain

def custom_preprocess(text):
    text = re.sub(r'\[([A-Z]+)\]', r'', text)
    text = re.sub(r'#(\S+)', r'\g<1>', text)
    text = re.sub(r'(\.)\1{2,}', '', text)
    text = re.sub(r'(\W)\1{2,}', r'\g<1>', text)
    
    return text

# Loading data
...

# Preprocess data
train_ds, test_ds, preproc = ktrain.text.texts_from_array(
    X_train, y_train, X_test, y_test,
    class_names=class_names, preprocess_fn=custom_preprocess
)
```

2. Let user change filter of `tf.keras.preprocessing.text.Tokenizer`

```py
def preprocess_train(self, train_text, y_train, verbose=1):
    ...
    # extract vocabulary
    self.tok = Tokenizer(num_words=self.max_features, filters=self.filters)
    self.tok.fit_on_texts(train_text)
    ...
```

## 4. Training plotting

ktrain currently only can plot value of [LR, loss and momentum](https://github.com/amaiya/ktrain/blob/v0.28.3/ktrain/core.py#L661). I would recommend to try plot other value if it's on training history key.

```py
else:
    if isinstance(plot_type, str) and plot_type not in self.history.history:
        raise ValueError('no lr in history: are you sure you used autofit or fit_onecycle to train?')
    plt.plot(self.history.history[plot_type])
    plt.title(f'{plot_type} Schedule')
    plt.ylabel(plot_type)

...
```

## 5. Use raw confidence score on QA module

ktrain currently perform softmax on confidence score. On few cases, this lead to first answer have very high score (> 0.9) while other answers have very low score. I would recommend to add parameter which doesn't perform softmax operation.

**Comment by amaiya** [comment]
Hi @ilos-vigil :  These all look like great suggestions to me, and PRs for these would be very welcome/appreciated.  For 1, it might be good to simply add a `train_val_names` argument to `texts_from_folder` and `images_from_folder`, but retain `train_test_names` for backwards compatibility for the time being.  If someone uses `train_test_names` instead of `train_val_names`, we can issue a deprecated warning perhaps. Thank you again!

**Comment by ilos-vigil** [comment]
Hi @amaiya,

* For point 1, i'd like to emphasize that naming inconsistency happened on many function/class and docstring. It'll take some effort to fix all inconsistencies.
* For point 2, how many EfficientNet model you'd like to support? There are 16 possible combination if you wish to support all of them (8 different EfficientNet model and random/imagenet weight).
* I'll make PR for point 4 and 5. As for point 3, i'll recheck ktrain source code before deciding whether to make PR or not.

**Comment by amaiya** [comment]
Point 1:  Sounds good.  Yes, we should use `val`  or validation set consistently.

Point 2:  I would tend to avoid overwhelming the user with too many choices, and perhaps select a couple of well-performing representative examples of the model.  In the paper, they seem to mention B7 (top accuracy, but most parameters) and B1 (slightly lower accuracy, but only 7.8M parameters), so perhaps we should choose these two models to start off with.

Thanks, @ilos-vigil .

**Comment by ilos-vigil** [comment]
Hi @amaiya, i'd like to mention that i won't be working on enchantment point 1-2. I mentioned it in case someone'd like to make PR for those point.

**Comment by amaiya** [comment]
Thanks, @ilos-vigil.   I think you meant 1 and 3, as 2 was implemented in PR 409.   Thanks again for the PRs you've implemented.    I'll make another release that includes them.

**Comment by ilos-vigil** [comment]
Yes, i meant point 1 and 3.
